tests/drivers/eeprom/src/main.c

```diff
@@ -164,50 +163,47 @@ static void test_zero_length_write(void)
 	zassert_equal(0, rc, "Unexpected error code (%d)", rc);
 }
 
-/* Run all of our tests on EEPROM device with the given label */
-static void run_tests_on_eeprom(const struct device *eeprom_dev, const struct device *i2c)
-{
-	zassert_true(device_is_ready(eeprom_dev), "EEPROM device is NOT ready");
-
-	k_object_access_grant(eeprom_dev, k_current_get());
-	eeprom = eeprom_dev;
-	/* if i2c is NULL, device_is_ready will report false */
-	if (device_is_ready(i2c)) {
-		/* If the test is using I2C, configure it */
-		k_object_access_grant(i2c, k_current_get());
-		i2c_configure(i2c, I2C_MODE_CONTROLLER |
-				I2C_SPEED_SET(I2C_SPEED_STANDARD));
-	}
+/* Get the device for an EEPROM devicetree alias */
+#define EEPROM_ALIAS_DEVICE_GET(alias) DEVICE_DT_GET(DT_ALIAS(alias))
 
-	ztest_test_suite(eeprom_api,
-			 ztest_user_unit_test(test_size),
-			 ztest_user_unit_test(test_out_of_bounds),
-			 ztest_user_unit_test(test_write_rewrite),
-			 ztest_user_unit_test(test_write_at_fixed_address),
-			 ztest_user_unit_test(test_write_byte),
-			 ztest_user_unit_test(test_write_at_increasing_address),
-			 ztest_user_unit_test(test_zero_length_write));
-	ztest_run_test_suite(eeprom_api);
-}
+/* Expand macro F for each EEPROM known devicetree alias */
+#define FOR_EACH_EEPROM_ALIAS(F) FOR_EACH(F, (), EEPROM_ALIASES)
 
-void test_main(void)
-{
-	const struct device *i2c_eeprom_0;
-#ifdef DT_N_ALIAS_eeprom_1
-	const struct device *i2c_eeprom_1;
-#endif
+/* Generate a test case function for an EEPROM devicetree alias */
+#define TEST_EEPROM_FUNCTION(alias, name)                                                          \
+	ZTEST_USER(eeprom, test_##alias##_##name) { test_##name(EEPROM_ALIAS_DEVICE_GET(alias)); }
+
+/* Generate test cases for an EEPROM devicetree alias */
+#define TEST_EEPROM(alias)                                                                         \
+	TEST_EEPROM_FUNCTION(alias, size)                                                          \
+	TEST_EEPROM_FUNCTION(alias, out_of_bounds)                                                 \
+	TEST_EEPROM_FUNCTION(alias, write_rewrite)                                                 \
+	TEST_EEPROM_FUNCTION(alias, write_at_fixed_address)                                        \
+	TEST_EEPROM_FUNCTION(alias, write_byte)                                                    \
+	TEST_EEPROM_FUNCTION(alias, write_at_increasing_address)                                   \
+	TEST_EEPROM_FUNCTION(alias, zero_length_write)
+
+/* Conditionally generate test cases for an EEPROM devicetree alias */
+#define TEST_EEPROM_COND(alias)                                                                    \
+	COND_CODE_1(DT_NODE_HAS_STATUS(DT_ALIAS(alias), okay), (TEST_EEPROM(alias)), ())
```

Comment:
This is super creative, I worry that this will be very hard to debug since assert failure line numbers won't map very well. I'm not a maintainer of this driver so I don't have an issue with it, just giving my 2 cents.

Reply:
+1

Reply:
Fair point. I will have to investigate how to do this in another way.